src/javascripts/modules/app.js

```diff
@@ -0,0 +1,58 @@
+/**
+ *  Example app
+ **/
+
+import I18n from '../../javascripts/lib/i18n'
+import { resizeContainer, render } from '../../javascripts/lib/helpers'
+import getDefaultTemplate from '../../templates/default'
+
+const MAX_HEIGHT = 1000
+const API_ENDPOINTS = {
+  organizations: '/api/v2/organizations.json'
+}
+
+class App {
+  constructor (client, appData) {
```

Comment:
Ok, so here goes 😆

I personally think passing client into the constructor is an anti-pattern for various reasons. It leads to harder to read code and gives the developer (the scaffold user) the wrong impression of how to use local/global/instance variables in my opinion.

As a starting point, the philosophy of object oriented programming (OOP). In my experience of writing code, but learning *more* by listening to other programmers on youtube and in lectures, and reading articles about programming, object oriented programming seems in my opinion often misunderstood and misused. I have come across plenty of code bases filled with `class` this and `object` that, that are inherently nothing more than `classified` (for lack of a better word) procedural code. Every article about OOP always has a simple example of vehicle and car, so I'm going to do the same thing 😆.

```
function drive (roads) {
  const road roads.shift()
  if (!road) return 'arrived!'
  return `we drove down ${road}`
}

class Vehicle {
  constructor (roads, color) {
    this.roads = roads
    this.color = color
    this.driverName = ''

    this.setup()
  }

  setup () {
    if (this.passengerSex === 'male') {
      this.driverName = 'Olaf'
    } else {
      this.driverName = 'Sara'
    }
  }

  drive () {
    drive(this.roads)
  }

  driver () {
    if (this.driverName)
      return `${this.driverName} is driving`
    else
      return 'Car is not driving'
  }
}
```

_Note: Maybe you're thinking "why put that drive function external, it should be part of the Vehicle.drive function", but I wanted to have an example external function, so bare with me._

This is OOP. We have a neat class, an external `helper` function, nothing wrong with that, we use `this.`, have a `constructor` and we could extend this `Vehicle` with `Car` if we wanted to. Everything dandy and we are writing OOP.

But what is OO really? Using `class`? Having a `constructor`? Setting variables to instances by using `this.`? What is more OO:

```
class A {
  constructor (data) {
    this._data = data
  }

  useData () {
    return this._data
  }
}
```

or

```
const _data

function setData (data)
  _data = data
}

function useData () {
  return _data
}
```

As far as the *consumer* of these files is concerned, *both* have a setData interface: `setData(123)` and `a = new A(123)`, both have an *internal* variable `const _data` and `this.data`, both have a useData function that uses that internal variable `useData` and `a.useData()`. And yet no class, no constructor, no instance variables. To top it off, I can actually *change* the `this.data` variable, because Javascript doesn't have private instance variables, which is a *downside* in this case, because as the "Class A" writer, i don't want the consumer to change the `this._data` variable directly. Hence the `_`. In other words, they do the same thing, except the class example is more vulnerable to bugs.

The point however was not dissing `Class`, the point was showing that *with* or *without* using `Class` I can write object oriented code.

Now we can go back to the Vehicle example, because just like we can write OO without `Class`, we can write non-OO (or procedural) code, whilst using `Class`. The Vehicle takes a color argument. That makes sense because Vehicles can *have* a color. But Vehicle can't have a road, because road isn't *owned* by Vehicle, road and Vehicle are seperate horizontally spaced concepts. They sit next to each other on a concept map. Merely because Vehicle *uses* roads, doesn't make it own road.

The second problem is that we have a `this.driverName`. As with `this.roads`, driverName is not owned by `Vehicle` but owned by `class Passenger`. If vehicle were to take an argument `passengers`, it would make more sense, but `driverName` is too specific to be linked to `Vehicle` directly. Further there is business logic on `driverName`. Lastely in the `setup` function and we are seperating `setting` a variable from `reading` a variable.

```
class B {
  s (data) {
    this.data = data

    if (this.data.indexOf('abc') !== -1) this.l()
  }

  l () {
    if (this.data === 'abc')
      this.other = 1
    else
      this.other = 2
  }

  o () {
    if (this.other === 2)
      this.other = 3
    else this.other = 1
  }

  r () {
    this.o()
    return this.other
  }
}
```

I know in these super simple examples it's blantely obvious what's going on, and maybe ask what's the problem? But the problem is that in order to *read* this code, the reader needs to have *full context* on code. Imagine a code base of 500 lines. Imagine the business logic sitting in a function that does multiple *other* things. Now imagine there is a bug because `r` doesn't return the correct data. Where to start? Looking at this code there is *no* way of knowing what `this.other` could possible have as a value. So multiple questions need to be answered before we start debugging.

1) Where does `this.other` get set?
  - In `l` and in `o`, but it's based on conditions which leads to:
2) When does `this.other` get set?
  - Depends on `this.data`, which gets set in `s`.
  - Depends on what `this.data` got set to.
  - Depends on if `l` got called.
  - Depends on if `o` got called.

So many questions, so many different *paths* for code to take, to read through. Business logic should be defined in it's *own* functions, and ideally not depended on call logic. What I mean by that is that in order to know what `this.other` is, I wouldn't need to see when `l` and `o` are called. I'm more tempted to run this code to see the behaviour. I feel as a writer (not as the reader) I have failed writing code that should be

1) easy to read, and
2) easy to maintain, and
3) easy to test for

3 is also difficult, because if I want to write tests for this, I need to test all possible code paths.

This is not OO code, but instead has become very procedural. In other words: What is OO about the following code?

```
function x (data) {
  let other

  if (data.indexOf('abc') !== -1) {
    if (data === 'abc')
      other = 1
    else
      other = 2
  }

  if (other === 2) other = 3
  else other = 1

  return other
}
```

Well that's actually not so bad at all. It takes an input, and it produces an output! This single `x` function is more readable than the code from `Class B`, because it doesn't change external state. Every time a function changes `external` state, it's harder to read. `Class B` changes external state in 3 functions (`s`, `l` and in `o`), but `x` doesn't ever.

It's not *always* possible to not change external state, but the state change should be obvious and in seperate functions.

```
class Dropdown {
  constructor (options) {
    this.options = options
  }

  addOption (option) {
    this.options.push(option)
  }

  removeOption (index) {
    this.options = this.options.slice(index, 1)
  }

  getOptions () {
    return this.options
  }
}
```

No issues here, although we change external state (options).
1) Dropdown *owns* options, like Vehicle *owns* color.
2) The `getOptions` isn't depended on `addOption` nor `removeOption`. Reading this code, you're probably not mystified about what `this.options` could possibly have as value.
3) I can tests all these functions very easily, and (this is the kicker) *independently*, because although I change state in multiple functions, they do *not* depend on each other. Unlike the `class B` example.

## Passing around variables
Object oriented code means that the object is a closed box with a defined door and window (interfaces). Nothing comes into the box without coming through the door, and leaves the box, without going trough the window.

Remember that helper function `drive` from the first class? Why is it passed a *internal* variable that no-one should touch? That passing of variables around violates OOP, because it doesn't follow the interface logic. That doesn't mean that internal variables can never be used as arguments with helper functions, but they should be copied by value (or cloned in case of object/array) so that the helper function, and this is the important thing, cannot change the *internal* state of the object.

Fun fact, we have a bug in ZAF because of exactly this problem.

```
   Box:         Other Box:
-------------     ------
|   ----    |     |    |
|   |  |<===|=====|=X  |
|   ----    |     ------
|           |
-------------
```

X should *not* have a reference to the inside of the box, because that goes against the OO concept. However `this.color` *can* be passed because it's copied by value.

Does everybody love my ascii art? :D

Since this is line 200, I'm gonna wrap it up.

- `Client` isn't *owned* by `class App`. Like road and vehicle, they are horizontally spaced concepts that sit next to each other.
- App shouldn't have *internal* variables that don't have anything to do with `App`.
- *internal* variables should not be passed to external functions/objects by reference.
- Business logic should be defined in it's *own* functions and not conditional (except for in the function itself)

I hope this clarifies some of the problems I have with client.

Reply:
While I don't disagree with most of what you're saying above with regards to OO, I don't necessarily agree that passing client to the constructor is completely flawed. I believe the design pattern we're using here is called [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection#Examples) and it has its merits. In our case, it allows us to create multiple classes that interact with the same client, this could be a different class per app location, or perhaps a different class per UI component of the app, all of which may need access to the same client. I think it's totally possible to avoid the problems you're referring to while still using this pattern.

The main goal we're trying to achieve here is to suggest a good basic structure for most apps. However, there's a limited amount of structure we can provide without knowing much about what the app needs to do.

What kind of basic structure would you suggest instead?

Reply:
Depends on the context, `roads` can also be a dependency of a `car`. A trialing auto pilot car needs to be set(created) with the `roads` it can drive on, and it may need to inform the `road` of it's presence(road.invoke('warning'))

So I don't think the `car` example above is appropriate to describe the `client` case here. `client` is tightly bond with the app, it has the app's location, user etc, it also is an "engine" that most `app` must have.

So an example of `roadsideAssistance` and `car` may make more sense.
```
roadsideAssistance = {
  rego: rego,
  callForHelp: function() {report(this.rego)}
}

class Car {
  constructor (roadsideAssistance, color, rego) {
    this.roadsideAssistance = roadsideAssistance
    this.color = color
    this.rego = rego
    this.driver = null
  }

  breakDown () {
    this.roadsideAssistance.callForHelp()
  }

  setDriver (person) {
    this.driver = person
  }

  getDriver () {
    return this.driver
  }
}
```

Reply:
> Client isn't owned by class App. Like road and vehicle, they are horizontally spaced concepts that sit next to each other.

True. But passing `Client` to App constructor doesn't necessarily mean `App` owns it, `App` owns the `this.client` variable which pointing to `Client`. 
**A car doesn't own the Factory/Brand/Roadside Assistance**, it's more of a relationship.

> App shouldn't have internal variables that don't have anything to do with App.

I'm not sure what internal variables are irrelevant with App?
Another fundamental concept of OOP is *encapsulation*. If App creation/function depends on a global object, you can't be sure the instance you created is solely depending on your creation params. Means `new Car('BMW', 'red')` is not necessarily same as another `new Car('BMW', 'red')`.

> internal variables should not be passed to external functions/objects by reference.

Internal variable can be a reference to an external object, this describes a relationship not necessarily ownership
```
const auFactory = new Factory('Melbourne, Australia')

const carA = new Car(auFactory)

function getFactoryAddress (factory) {
  return car.factory.address
}

getFactoryAddress(carA.factory)  ==> 'Melbourne, Australia'

auFactory.relocate('Adelaide, Australia')

getFactoryAddress(carA.factory) ==> 'Adelaide, Australia'
```

> Business logic should be defined in it's own functions and not conditional (except for in the function itself)

This relates to anything particular in the scaffold?

Reply:
> I believe the design pattern we're using here is called dependency injection and it has its merits.

In my opinion we are using the dependency injection technique wrong, and not how it is should to be used. I feel that `class App` is thought of, as representing app code, but that is not the case in my mind. *All* code is the app code. In other words, currently we are conflating the `App` class to be the same as the app code.

This is not true because `App` does not represent the code running in the iframe, it is only a portion of the code. There are a lot of other _things_ (lack of better word) going on, that are app code, but do not sit in the `App` class.

The perfect example of that is resizing the app. This code does not sit in the `App` class, but is in my opinion surely part of the app code. This is where I think the problem arises, since we currently having a disconnect between `App` code and actual app code. Because this disconnect has happened, we need to pass _internal_ variables to other app code, although they are not part of `App` code (but indeed part of app code)

The first line of that wiki page states

> In software engineering, dependency injection is a technique whereby one object (or static method) supplies *the dependencies of another object*

(emphasis mine). This is not what happens. The current dependency injection supplies the dependency for *multiple* other objects, which are not nested inside another, but sit next to each other. Because it is used wrong in my opinion, we need to pass internal variables around, which makes for hard to read, hard to maintain and hard to change code. It also leads to _more_ code, because we need mechanisms in place to handle this _passing around of variables_.

If client were *only* used in `App`, that would make sense, but that is not the case, we are exporting `Client` to other parts because we need to.

However there is a good solution that we already use! Simply treat client like I18n. Both of them are accessible globally, we do not use dependency injection to pass I18n around, we simply see it as another tool in global context. Just like we should see the client as another tool in global context.

`lib/client.js`
```
export default ZAFClient.init()
```

another other file that needs client `code.js`
```
import client from './lib/client'
```

Reply:
> But passing Client to App constructor doesn't necessarily mean App owns it, App owns the this.client variable which pointing to Client.

I see your argument about `App` not _owning_ client, and maybe owning wasn't the correct wording on my side. `App` can certainly have a `client`, the main problem is that this is passed around to other objects/functions.

> to anything particular in the scaffold?

No, sorry 😆. I have seen this a lot in apps though, so I thought it was interesting to point out.

I like your last example about `auFactory` because it shows some of the thinking going on around `App` and `helpers`. In my opinion it should be

```
class Factory
  constrcutor (address) {
    this.address = address
  }

  getFactoryAddress (factory) {
    return this.address
  }
}

const auFactory = new Factory('Melbourne, Australia')
auFactory.getFactoryAddress() ==> 'Melbourne, Australia'
auFactory.relocate('Adelaide, Australia')
auFactory.getFactoryAddress() ==> 'Adelaide, Australia'
```